fix: correctly resolve dependencies for CT onboarding when using Yarn Plug n Play

lmiller1990 opened the pull request:
<!-- Thanks for contributing! PLEASE...
- Read our contributing guidelines: https://github.com/cypress-io/cypress/blob/develop/CONTRIBUTING.md 
- Read our Code Review Checklist on coding standards and what needs to be done before a PR can be merged: https://github.com/cypress-io/cypress/blob/develop/CONTRIBUTING.md#Code-Review-Checklist
- Mark this PR as "Draft" if it is not ready for review.
- Make sure you set the correct base branch based on what packages you're changing: https://github.com/cypress-io/cypress/blob/develop/CONTRIBUTING.md#branches
-->

- Closes https://github.com/cypress-io/cypress/issues/25960

### Additional details
<!-- Examples:
- Why was this change necessary?
- What is affected by this change?
- Any implementation details to explain?
-->

Correctly detect installed modules for projects using Yarn 3 and Plug n Play.

### Steps to test
<!--
For non-trivial behavior changes, list the steps that a reviewer should follow to validate the new behavior.
This is not meant to be the only testing performed by a reviewer, just the "happy path" that leads to the new behavior.
-->

Make a Yarn 3 project with Vue and Vite (or whatever deps you like).

```sh
mkdir yarn-3-repro
cd yarn-3-repro
yarn set version berry
yarn init -y
yarn add vite vue cypress                                                                  
```

Try setup CT with the current cypress (just do `yarn cypress open`). It does not detect your deps.

Now use this branch - it detects them correctly! 🎉 

### How has the user experience changed?
<!-- Provide before and after examples of the change.
Screenshots or GIFs are preferred. -->

Using Yarn 3 to onboard does not give the impression something is broken.

### PR Tasks
<!-- 
These tasks must be completed before a PR is merged.
If a task does not apply, write [na] instead of checking the box.
DO NOT DELETE the PR checklist.
-->

- [x] Have tests been added/updated?
- [ ] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? <!-- Link to PR here -->
- [ ] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

lmiller1990 committed:
patch resolve package and use corret path for Yarn PnP module resolution

lmiller1990 committed:
add test

lmiller1990 commented on packages/scaffold-config/src/frameworks.ts:
```diff
@@ -14,10 +14,42 @@ export type WizardBundler = typeof dependencies.WIZARD_BUNDLERS[number]
+    // using Yarn PnP. You cannot `fs.readJson`, since the module is zipped.
+    // use require.resolve - The PnP runtime (.pnp.cjs) automatically patches Node's
+    // fs module to add support for accessing files inside Zip archives.
+    // @see https://yarnpkg.com/features/pnp#packages-are-stored-inside-zip-archives-how-can-i-access-their-files
```
Might be worth reading https://yarnpkg.com/features/pnp for more context.

lmiller1990 commented on packages/scaffold-config/src/frameworks.ts:
```diff
@@ -14,10 +14,42 @@ export type WizardBundler = typeof dependencies.WIZARD_BUNDLERS[number]
+    // we only need to register this once, when the project check dependencies for the first time.
+    if (yarnPnpRegistrationPath !== projectPath) {
+      try {
+        const pnpapi = require(path.join(projectPath, '.pnp.cjs'))
```
`.pnp.cjs` is regenerated by Yarn every time you install a module. We only want to require it once, since the trick Yarn PnP uses is to dynamically monkeypatch `module` and do some other weird hacks to facilitate the various features of PnP. That is what `setup()` does. 

It's a bit of a hack (Yarn 3, I mean, not what I'm doing here - this is how they recommend you load the `.pnp.cjs` if you are doing it manually). Yarn 3 does run really fast, though, because of this... although Yarn 1 -> Yarn 2.x + PnP is a heck of a lift for a lot of code bases and projects, which is why many people are stuck on v1 (including FB, the (original) authors of Yarn).

It wouldn't matter if we executed this block it every time - it would just be a bit of pointless overhead. It would be next to negligible, since Node.js caches `require` calls anyway, but this is cleaner. We *do* need to execute it once for each new project (think global mode - you can switch projects, each having a different `.pnp.cjs`).

lmiller1990 commented on packages/scaffold-config/src/frameworks.ts:
```diff
@@ -14,10 +14,42 @@ export type WizardBundler = typeof dependencies.WIZARD_BUNDLERS[number]
     debug('detecting %s in %s', dependency.package, projectPath)
 
+    // we only need to register this once, when the project check dependencies for the first time.
+    if (yarnPnpRegistrationPath !== projectPath) {
```
So the reason I compare the paths here if so if you are using global mode, it'll correctly try to require the pnp.cjs once each time you change projects.

lmiller1990 commented on patches/resolve-package-path+4.0.3.patch:
```diff
@@ -0,0 +1,14 @@
+             // current release. This is due to Yarn 1.13 and resolve interoperating
+             // together seamlessly.
++            if (!pnp) {
++                pnp = require(require.resolve('pnpapi', { paths: [baseDir] }))
```
I send the author a PR https://github.com/stefanpenner/resolve-package-path/issues/62

Not sure if he's interested, I think patching this is okay for now. We need to do this, otherwise it won't look in the right directroy for the pnpapi, which is a virtual module dynamically created when you do `require(pnp.cjs).setup()`. Figuring this whole thing out was pretty tedious, I am glad it's done.

lmiller1990 committed:
fix logic

lmiller1990 commented on patches/resolve-package-path+4.0.3.patch:
```diff
@@ -0,0 +1,19 @@
+diff --git a/node_modules/resolve-package-path/lib/index.js b/node_modules/resolve-package-path/lib/index.js
+index a1463f7..00d83b8 100644
+--- a/node_modules/resolve-package-path/lib/index.js
++++ b/node_modules/resolve-package-path/lib/index.js
```
Sent the author a patch https://github.com/stefanpenner/resolve-package-path/issues/62

lmiller1990 replied:
Problem is the module is installed in the same directory as the project, which in our case it isn't - the module is in the binary.

lmiller1990 committed:
changelog

lmiller1990 committed:
log

mike-plummer commented on packages/scaffold-config/src/frameworks.ts:
```diff
@@ -14,10 +14,42 @@ export type WizardBundler = typeof dependencies.WIZARD_BUNDLERS[number]
+    // NOTE: this *must* be required **after** the call to `pnpapi.setup()`
+    // or the pnpapi module that is added at runtime by Yarn PnP will not be correctly used
+    // for module resolution.
+    const resolvePackagePath = require('resolve-package-path')
```
omg, _never_ would have figured this one out. Very nice!

mike-plummer commented on packages/scaffold-config/src/frameworks.ts:
```diff
@@ -14,10 +14,42 @@ export type WizardBundler = typeof dependencies.WIZARD_BUNDLERS[number]
 
 export type CodeGenFramework = Cypress.ResolvedComponentFrameworkDefinition['codeGenFramework']
 
+const yarnPnpRegistrationPath = new Map<string, { usesYarnPnP: boolean }>()
```
Very minor, but we could simplify this map to just be a `Map<string, boolean>`. Do you foresee any other fields being tracked that would require an object value?

lmiller1990 replied:
I had this thought too, my reasoning was the property `usesYarnPnP` makes it more obvious what this is used for. Could go either way, I wish TS had a feature where you could label the value, eg `Map<string, usesPnP as boolean>` or something...

mike-plummer commented on packages/scaffold-config/src/frameworks.ts:
```diff
@@ -14,10 +14,42 @@ export type WizardBundler = typeof dependencies.WIZARD_BUNDLERS[number]
+    return require(require.resolve(packageFilePath))
+  }
+
+  return await fs.readJson(packageFilePath)
```
Maybe I'm missing an edge case, but couldn't we just use the `require(require.resolve(packageFilePath))` logic in both cases? PnP handles if the module is zipped, and if we aren't in PnP it will be a flat file that should be normally require-able

lmiller1990 replied:
I think there was an edge case where this doesn't work, let me double check.

mike-plummer commented on packages/scaffold-config/src/frameworks.ts:
```diff
@@ -14,10 +14,42 @@ export type WizardBundler = typeof dependencies.WIZARD_BUNDLERS[number]
+    // we only need to register this once, when the project check dependencies for the first time.
+    if (!yarnPnpRegistrationPath.has(projectPath)) {
+      try {
+        const pnpapi = require(path.join(projectPath, '.pnp.cjs'))
```
What about a scenario where a Cypress project is nested into a subdirectory of a yarn project? Do we need to traverse upwards to find the first `.pnp.cjs` we encounter?

Theres a method on the [`module` module](https://yarnpkg.com/advanced/pnpapi#requiremodule) that seems to do this for us in a PnP context (find `.pnp.cjs`, load, and return `pnpapi` (which we can ignore))

lmiller1990 replied:
Ah, forgot the nested edge case, I will verify what happens and address using this if required. Yarn has a ton of APIs for this, as you pointed out, we can probably use one.

lmiller1990 replied:
Painful, but for this API to exist you need to be in a PnP context... which you get by requiring the file in the first place 🤦 

I think we need to manually traverse up, lucky we've already got some code that does that, `isRepositoryRoot`.

mike-plummer reviewed (approved):
Amazing job tracking this down 🤯 

I have just a couple style comments and one potential edge case, but none of them are real blockers and can be addressed as follow-ons if necessary so this can make the release

emilyrohrbough commented on cli/CHANGELOG.md:
```diff
@@ -8,6 +8,7 @@ _Released 04/11/2023 (PENDING)_
  - Capture the [Azure](https://azure.microsoft.com/) CI provider's environment variable [`SYSTEM_PULLREQUEST_PULLREQUESTNUMBER`](https://learn.microsoft.com/en-us/azure/devops/pipelines/build/variables?view=azure-devops&tabs=yaml#system-variables-devops-services) to display the linked PR number in the Cloud. Addressed in [#26215](https://github.com/cypress-io/cypress/pull/26215).
  - Fixed an issue in the onboarding wizard where project framework & bundler would not be auto-detected when opening directly into component testing mode using the `--component` CLI flag. Fixes [#22777](https://github.com/cypress-io/cypress/issues/22777).
  - Fix an edge case in Component Testing where a custom `baseUrl` in `tsconfig.json` for Next.js 13.2.0+ is not respected. This was partially fixed in [#26005](https://github.com/cypress-io/cypress/pull/26005), but an edge case was missed. Fixes [#25951](https://github.com/cypress-io/cypress/issues/25951).
+ - Correctly detect and resolve dependencies when configuring Component Testing in projects using Yarn's Plug n Play feature. Fixes [#25960](https://github.com/cypress-io/cypress/issues/25960).
```
Is this a Yarn 3 specific this? the test calls it out as yarn 3

mike-plummer replied:
@emilyrohrbough The original issue here was written up against Yarn 3, but Yarn added PnP in ~1.12 as an opt-in capability. Very unlikely for a 1.x project to be using it, but it's possible

emilyrohrbough replied:
gottchya, is there a link to this feature we can add for reference?

mike-plummer replied:
@emilyrohrbough Was actually just thinking about that - there's separate docs for Yarn 1.x and 3.x; I suppose we would be better off linking to the newest stuff

emilyrohrbough replied:
yeah I'd prob just do the newer flavor of it

mike-plummer committed:
Add link to pnp docs

mike-plummer committed:
Merge branch 'develop' into issue-25960

lmiller1990 committed:
Merge remote-tracking branch 'origin/develop' into issue-25960

lmiller1990 committed:
recursively search upwards for pnp.cjs

lmiller1990 committed:
use require.resolve no matter what

lmiller1990 commented:
@mike-plummer 

1. I addressed the "find pnp.cjs recursively in monorepo" comment, this commit => https://github.com/cypress-io/cypress/pull/26452/commits/775314a9556bfe69717b9540a0d15c59195d25a0
2. I addressed "can't we just `require.resolve` no matter what here => https://github.com/cypress-io/cypress/pull/26452/commits/47b41bd4aadda7b3b804bac0819c62d3e42e4a70

@astone123 can you take a look? According to https://github.com/cypress-io/cypress/pull/26033 "There are some dependencies that we can't resolve with [require.resolve](https://nodejs.org/api/modules.html#requireresolverequest-options), because of one of the following reasons:"... but this seems to not be the case anymore, I'm using `require.resolve` on all the things and they all pass. I wonder if `resolve-package-path` does not have the same limitations?

astone123 commented:
@lmiller1990 `resolve-package-path` does **not** have those same limitations, which is why I introduced it in https://github.com/cypress-io/cypress/pull/26033. 

I think that [this](https://github.com/cypress-io/cypress/pull/26452/files#diff-dce5144c067a639808029e38cef9e4e90a64dec50aeec74e6064f3cb31a56981R19) will work around those limitations though because the difference here is that we've used `resolve-package-path` to get the full path of the package file, rather than trying to resolve with just the module name.

The issue before was that we were trying to do 
```js
require.resolve('my-module')
```
which will only work if the package file either doesn't specify any `exports`, or does **and includes** `package.json`. But in this case we'd be doing
```js
require.resolve('path/to/my-module/package.json')
```
so we shouldn't have issues

lmiller1990 committed:
Merge branch 'develop' into issue-25960

sfsepark commented:
@lmiller1990 
Unfortunately, even after 12.10.0 deployment, there is still an error in putting the reporter in the settings in the pnp environment. I thought it would be solved after this patch is applied, but it's very unfortunate.

Is there a solution to this? Still, is there a solution other than changing the cypress internal code calling the reporter?

ref issue : https://github.com/cypress-io/cypress/issues/18922

```
Error loading the reporter: cypress-multi-reporter

We searched for the reporter in these paths:

 - /Users/user/Documents/GitHub/.../project/cypress-multi-reporter
 - /Users/user/Documents/GitHub/.../project/node_modules/cypress-multi-reporter

Learn more at https://on.cypress.io/reporters

Error: Cannot find module '/Users/user/Documents/GitHub/.../project/node_modules/cypress-multi-reporter'
Require stack:
- /Users/user/Library/Caches/Cypress/12.10.0/Cypress.app/Contents/Resources/app/packages/server/lib/reporter.js
    at Module._resolveFilename (node:internal/modules/cjs/loader:940:15)
    at n._resolveFilename (node:electron/js2c/browser_init:249:1105)
    at PackherdModuleLoader._tryResolveFilename (evalmachine.<anonymous>:1:737331)
```

lmiller1990 commented:
@sfsepark I'll reply in #18922

chrisbreiding commented:
Looks like this may have caused a regression: https://github.com/cypress-io/cypress/issues/26676

lmiller1990 commented:
@chrisbreiding I think you are right. We might be best to simply revert that commit until we've got a fix - I took a quick look at the new issue, #26676, but I don't think it's a straight forward fix.
